x-pack/agent/pkg/agent/application/filters/constraints_filter.go

```diff
@@ -0,0 +1,239 @@
+	validateVersionFuncName = "validate_version"
+)
+
+// List of variables available to be used in constraint definitions.
+const (
+	// `agent.id` is a generated (in standalone) or assigned (in fleet) agent identifier.
+	agentIDKey = "agent.id"
+	// `agent.version` specifies current version of an agent.
+	agentVersionKey = "agent.version"
+	// `host.architecture` defines architecture of a host (e.g. x86_64, arm, ppc, mips).
+	hostArchKey = "host.architecture"
+	// `os.family` defines a family of underlying operating system (e.g. redhat, debian, freebsd, windows).
+	osFamilyKey = "os.family"
+	// `os.kernel` specifies current version of a kernel in a semver format.
+	osKernelKey = "os.kernel"
+	// `os.platform` specifies platform agent is running on (e.g. centos, ubuntu, windows).
+	osPlatformKey = "os.platform"
+	// `os.version` specifies version of underlying operating system (e.g. 10.12.6).
+	osVersionKey = "os.version"
+)
+
+var (
+	boolexpVarStore    *constraintVarStore
+	boolexpMethodsRegs *boolexp.MethodsReg
+)
+
+// ConstraintFilter filters ast based on included constraints.
+func ConstraintFilter(log *logger.Logger, ast *transpiler.AST) error {
+	// get datasources
+	dsNode, found := transpiler.Lookup(ast, datasourcesKey)
+	if !found {
+		return nil
+	}
+
+	dsListNode, ok := dsNode.Value().(*transpiler.List)
+	if !ok {
+		return nil
+	}
+
+	dsList, ok := dsListNode.Value().([]transpiler.Node)
+	if !ok {
+		return nil
+	}
+
+	// for each datasource
+	i := 0
+	originalLen := len(dsList)
+	for i < len(dsList) {
+		constraintMatch, err := evaluateConstraints(log, dsList[i])
+		if err != nil {
+			return err
+		}
+
+		if constraintMatch {
+			i++
+			continue
+		}
+		dsList = append(dsList[:i], dsList[i+1:]...)
+	}
+
+	if len(dsList) == originalLen {
+		return nil
+	}
+
+	// Replace datasources with limited set
+	if err := transpiler.RemoveKey(datasourcesKey).Apply(ast); err != nil {
+		return err
+	}
+
+	newList := transpiler.NewList(dsList)
+	return transpiler.Insert(ast, newList, datasourcesKey)
+}
+
+func evaluateConstraints(log *logger.Logger, datasourceNode transpiler.Node) (bool, error) {
+	constraintsNode, found := datasourceNode.Find(constraintsKey)
+	if !found {
+		return true, nil
+	}
+
+	constraintsListNode, ok := constraintsNode.Value().(*transpiler.List)
+	if !ok {
+		return false, errors.New("constraints not a list", errors.TypeConfig)
+	}
+
+	constraintsList, ok := constraintsListNode.Value().([]transpiler.Node)
+	if !ok {
+		return false, errors.New("constraints not a list", errors.TypeConfig)
+	}
+
+	for _, c := range constraintsList {
+		strval, ok := c.(*transpiler.StrVal)
+		if !ok {
+			return false, errors.New("constraints is not a string")
+		}
+
+		constraint := strval.String()
+		if isOK, err := evaluateConstraint(constraint); !isOK || err != nil {
+			if err == nil {
+				// log only constraint not matching
+				log.Infof("constraint '%s' not matching for datasource: %v", constraint, datasourceNode.String())
```

Comment:
We might want to reduce what we want to output to the log to the type of the **datasource** and **id**? This is just to make sure we do not expose credentials in the log this could be a problem with metricbeat where datasource have a higher change of containing secrets.

Reply:
constriant is at datasource level and as such it does not contain type, ID is optional for now. i think namespace, output is a identifying configuration and id if present will be logged

Reply:
sound good ^